probe_src/probe_py/analysis.py

```diff
@@ -58,25 +59,53 @@ def last(pid: int, exid: int, tid: int) -> Node:
             return exit_node
         else:
             return proc_to_ops[(pid, exid, tid)][-1]
+        
+    def get_first_pthread(pid, exid, target_pthread_id):
+        for kthread_id, thread in process_tree_prov_log.processes[pid].exec_epochs[exid].threads.items():          
+            op_index = 0
+            for op in thread.ops:               
+                if op.pthread_id == target_pthread_id:
+                    return (pid, exid, kthread_id, op_index)
+                op_index+=1  
+        return (pid, -1, target_pthread_id, -1) 
+
+    def get_last_pthread(pid, exid, target_pthread_id):
+        for kthread_id, thread in process_tree_prov_log.processes[pid].exec_epochs[exid].threads.items():          
+            op_index = len(thread.ops) - 1
+            ops = thread.ops
+            while op_index >= 0:  
+                op = ops[op_index]             
+                if op.pthread_id == target_pthread_id:
+                    return (pid, exid, kthread_id, op_index)
+                op_index-=1
+        return (pid, -1, target_pthread_id, -1)
 
     # Hook up forks/joins
     for node in list(nodes):
         pid, exid, tid, op_index = node
         op = process_tree_prov_log.processes[pid].exec_epochs[exid].threads[tid].ops[op_index].data
+        global target
         if isinstance(op, CloneOp):
-            if op.flags & CLONE_THREAD:
+            if op.task_type == TaskType.TASK_PID:
                 # Spawning a thread links to the current PID and exec epoch
-                target = (pid, exid, op.child_thread_id)
+                target = (pid, exid, op.task_id)
+            if op.task_type == TaskType.TASK_PTHREAD:
+                target_pthread_id = op.task_id
+                dest = get_first_pthread(pid, exid, target_pthread_id)
```

Comment:
What happens if the thread does not contain any ops of the given pthread id?

Reply:
we an edge from the `CloneOp` to `(pid, -1, target_pthread_id, -1)` , I added a check that if a `cloneOp` has an edge to such node it's reported that no ops were found.

Reply:
Not every kernel thread will have an Op of the relevant user-thread (pthread or ISO C thread). We should just ignore the case when there is no "match" in this kthread.